### PR TITLE
dispatch to osemgrep by default

### PR DESCRIPTION
**Pull request by aryx**
... but quickly dispatch back to pysemgrep for most commands,
unless you used --experimental.

But 'semgrep interactive' does not require anymore the --experimental!
Next target is 'semgrep lsp' e2e in OCaml, without --experimental.

test plan:
cd cli
pipenv install --dev
pipenv shell
time semgrep --help
=> 0.090s
time pysemgrep --help
=> 0.368s
semgrep --experimental scan --help
=> osemgrep man page
semgrep scan --help
=> pysemgrep man page as we default back to pysemgrep
semgrep interactive -l ocaml .
=> works! without --experimental

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by aryx** [comment]
@brandonspark you can now simply do 'semgrep interactive -l ...', no need --experimental anymore.

**Comment by brandonspark** [comment]
wooh!

File: cli/bin/semgrep
Comment (ajbt200128) [inline cli/bin/semgrep:90]: I think that a lot of the errors in the tests are because we're not passing environment variables here, which is important to not break semgrep in CI.

File: cli/bin/semgrep
Comment (ajbt200128) [inline cli/bin/semgrep]: No, we should not call execvp here, it will break the fix we just made for tests + autocomplete. I tried adding a longer explanation in that PR but the short answer is calling execvp from python to run more python breaks things

File: cli/bin/semgrep
Comment (ajbt200128) [inline cli/bin/semgrep:110]: We should make sure we add this to the documentation right away, so users know how to get pysemgrep if things start regressing

File: cli/bin/semgrep
Comment (ajbt200128) [reply cli/bin/semgrep:90]: Actually checking the docs, it seems execvp will inherit environment variables by default so I think this is fine

File: src/osemgrep/cli/CLI.ml
Comment (ajbt200128) [inline src/osemgrep/cli/CLI.ml]: Can we make this a better error message than assert false since this is now exposed to end users?

**Review by ajbt200128 — COMMENTED** [review]
I know I said I would try and switch the python tests to calling semgrep directly, instead of through the pylib, but I feel that change might make sense in this PR (or before). I worry that there is some edge case (something in CI that requires environment variables or something like that) that we're not going to catch without testing by calling it this way.

I also wonder if maybe it's worth making the CLI.ml changes (i.e. osemgrep falls back to pysemgrep), and trying to dog food first by switching everywhere we use semgrep to `semgrep --experimental`.

**Comment by brandonspark** [comment]
seems fine to me, i'll wait on another's approval though

File: cli/bin/semgrep
Comment (aryx) [reply cli/bin/semgrep:90]: yes, children inherits the environment of the parent (UNIX got many things right!)

**Comment by aryx** [comment]
Good comments @ajbt200128 . I agree getting e2e tests to use semgrep is high-priority!
Also here we very quickly dispatch back to pysemgrep for everything (except interactive, which anyway is not supported in pysemgrep), so I think we should be fine.

**Comment by aryx** [comment]
@ajbt200128 any idea why the test_ls.py tests are now failing?

File: cli/bin/semgrep
Comment (aryx) [reply cli/bin/semgrep]: I've added a bigger comment

File: src/osemgrep/cli/CLI.ml
Comment (aryx) [reply src/osemgrep/cli/CLI.ml]: no it should never happen, but I can make an error message (even though it should never happen).

File: cli/bin/semgrep
Comment (aryx) [reply cli/bin/semgrep:110]: I've added a changelog entry.

File: cli/bin/semgrep
Comment (emjin) [inline cli/bin/semgrep]:
```suggestion
# In the future we may have different entry points when packaging Semgrep
```

File: cli/bin/semgrep
Comment (emjin) [inline cli/bin/semgrep]:
```suggestion
# explicitly as in /path/to/somewhere/bin/semgrep, but this means
```

File: src/osemgrep/cli/CLI.ml
Comment (ajbt200128) [inline src/osemgrep/cli/CLI.ml:165]: Ahh ok, tests are failing for the language server because before we would run pysemgrep -> call `osemgrep lsp` and use it as a subprocess. now this never happens, I'll add a fix

**Comment by IagoAbal** [comment]
How much does this add to the start-up time of `semgrep` ?

File: cli/bin/semgrep
Comment (emjin) [inline cli/bin/semgrep]: For my understanding: why does this not need `sys.exit`?

File: cli/src/semgrep/lsp/server.py
Comment (emjin) [inline cli/src/semgrep/lsp/server.py]: Why does this need `--experimental`?

**Comment by aryx** [comment]
I would say: 0.090 
 - 0.15 for the Python startup time of cli/bin/semgrep (does not use much import so it's fast)
 - 0.75 for the OCaml startup time

File: cli/src/semgrep/lsp/server.py
Comment (aryx) [inline cli/src/semgrep/lsp/server.py]: This need to be "semgrep", not "osemgrep", because osemgrep is not part of the pip package nor Docker image.

File: src/osemgrep/cli_scan/Scan_CLI.ml
Comment (emjin) [inline src/osemgrep/cli_scan/Scan_CLI.ml:767]: Why does this need to error for MExperimental?

File: cli/src/semgrep/lsp/server.py
Comment (aryx) [reply cli/src/semgrep/lsp/server.py]: In fact I don't even know how this can work ..., what is the path of this osemgrep and how our test infrastructure can find this osemgrep.

File: src/osemgrep/cli_scan/Scan_CLI.ml
Comment (emjin) [reply src/osemgrep/cli_scan/Scan_CLI.ml:767]: Edit: I see. Maybe we could have a better name, e.g. MLegacyCompatible, or we could do something like MExperimental, MLegacy, MPySemgrep

File: cli/src/semgrep/lsp/server.py
Comment (ajbt200128) [reply cli/src/semgrep/lsp/server.py]: we're not calling osemgrep, we're calling semgrep-core and passing the above as arguments. I'll rename `cmd` to `args`. This is functioning the same as `call_osemgrep`, but looks different because we're using Popen and not os.exec here so we can keep it as a child process

File: cli/src/semgrep/lsp/server.py
Comment (emjin) [inline cli/src/semgrep/lsp/server.py:36]: Why do we need to add `--experimental`?

Comment (emjin) [reply cli/src/semgrep/lsp/server.py:36]: I now understand why; would prefer better names for these flags

**Review by emjin — APPROVED** [review]
It looks good to me; I don't know much about the exec-ing but the tests pass. I would really prefer better names for the flags `--develop`, `--experimental`, `--legacy`, however. `--experimental` does not do what I intuitively expect

File: cli/bin/semgrep
Comment (ajbt200128) [reply cli/bin/semgrep]: os.exec is a thin wrapper to the OS's exec, which starts the process in place of the caller.

Comment (ajbt200128) [reply cli/bin/semgrep]: i.e. not as a child process, it just replaces it

**Comment by emjin** [comment]
> Also here we very quickly dispatch back to pysemgrep for everything (except interactive, which anyway is not supported in pysemgrep), so I think we should be fine.

I think the question here is about whether the tests are indeed calling this path, or if they're directly invoking pysemgrep and skipping this exec flow. If it's the latter, I'm not as comfortable merging without adding some tests specifically for this flow

File: cli/src/semgrep/lsp/server.py
Comment (ajbt200128) [reply cli/src/semgrep/lsp/server.py]: before, calling `osemgrep subcommand` would just run osemgrep's implementation of that subcommand. Now, osemgrep calls back to pysemgrep by default, and only runs it's own implementation of a subcommand if --experimental is passed. Once I port login stuff, we will be able to delete this whole file and just use osemgrep directly

**Comment by ajbt200128** [comment]
@emjin yea right now the tests call the invoke semgrep function on the python side. None of the tests handle anything with the execution flow and I have the same worries hence the comment. Note that messing with the entrypoint stuff previously is what broke the test subcommand and autocomplete, as it wasn't caught by tests.

**Review by emjin — CHANGES_REQUESTED** [review]
Did not realize this didn't go through automated tests; we should have at least one test for this entry point

**Comment by aryx** [comment]
Which test do you propose?

**Comment by emjin** [comment]
Can we set one up that calls semgrep and doesn't dispatch directly to the Python? From there I don't care what it does, a basic rule on basic file is fine

**Comment by aryx** [comment]
Making our e2e test switch from calling semgrep via internal python function to forking semgrep is the right long-term solution, but will require a different PR like I said to Austin above. This will also slow down quite a lot 'make test' (which I'm fine with; better to be correct than fast but incorrect).

**Comment by aryx** [comment]
And why this new e2e test would give you more confidence?
I can start an e2e test infra in ocaml (instead of pytest), but since most of our e2e tests are using pytest (and calling invoke_semgrep() direcly), not sure one ocaml e2e test will improve things a lot.

**Comment by aryx** [comment]
BTW a few of our CI tests were failing when I made this PR, meaning there are few places where we exercise
this cli/bin/semgrep binary (including test_ls.py, our Dockerfile, etc.).

**Review by emjin — APPROVED** [review]
Ok, I did not have clarity on how the semgrep tests work, so I went and checked.

In semgrep_runner.py:
```
# The semgrep command suitable to run semgrep as a separate process.
# It's something like ["semgrep"] or ["python3"; -m; "semgrep"] or
# ["/path/to/osemgrep"].
SEMGREP_BASE_COMMAND: List[str] = (
    [OSEMGREP_PATH] if USE_OSEMGREP else [sys.executable, "-m", "semgrep"]
)

SEMGREP_BASE_COMMAND_STR: str = " ".join(SEMGREP_BASE_COMMAND)
```

This looks to me like it's executing `semgrep` (or `osemgrep`), and it's used in `test_check`, including at least one matching test. That satisfies what I was asking for.

**Comment by ajbt200128** [comment]
Oh good catch, I didn't know test_check calls semgrep directly in some of the tests

**Comment by ajbt200128** [comment]
Actually it does not, `python -m semgrep` is different it imports the python module, and does not call the entrypoint :(

**Comment by ajbt200128** [comment]
Just confirmed, breaking osemgrep does not break any python tests :( (as suggested by emma, I changed the pysemgrep callback in CLI to something random)

**Comment by IagoAbal** [comment]
I am not sure why but running benchmarks to compare develop vs this PR, as well as 1.33.2 vs this PR, I get one more finding with this PR when analyzing `gitlabhq`:

```diff
 ┌───────────────────┐
-│ 113 Code Findings │
+│ 114 Code Findings │
 └───────────────────┘
+                                                                                
+    repos/gitlabhq/app/assets/javascripts/issues/create_merge_request_dropdown.j
+  s                                                                             
+       javascript.browser.security.open-redirect-from-function.js-open-redirect-
+  from-function                                                                 
+          The application accepts potentially user-controlled input `{ data }`  
+  which can control the                                                         
+          location of the current window context. This can lead two types of    
+  vulnerabilities open-                                                         
+          redirection and Cross-Site-Scripting (XSS) with JavaScript URIs. It is
+  recommended to                                                                
+          validate user-controllable input before allowing it to control the    
+  redirection.                                                                  
+          Details: https://sg.run/P39e                                          
+                                                                                
+          161┆ window.location.href = data.url;
```

Could it be because this PR needs to be rebased ?

**Comment by ajbt200128** [comment]
Ok, I updated `semgrep_runner.py` to actually call the correct entrypoint. So it should work how Emma expects now, and I confirmed breaking osemgrep breaks some tests in `test_check.py`

**Comment by ajbt200128** [comment]
Ok I think the failing tests are now due to the actual changes in this PR

**Comment by ajbt200128** [comment]
It looks like those tests were failing because path would be empty, and so python couldn't find the semgrep module when running them. I think it's fair to assume that python's module directory will be on the path, as is the case with literally every unix computer that has python installed

**Comment by aryx** [comment]
good investigation @emjin. TIL we were already doing real e2e testing in test_check.py

**Comment by IagoAbal** [comment]
False alarm, same number of findings now. 👍

**Pull request merged**
